debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ParallelSnapshotReader.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mysql;
+
+import io.debezium.config.Configuration;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+
+/**
+ * A reader that runs a {@link ChainedReader} consisting of a {@link SnapshotReader} and a {@link BinlogReader}
+ * for all tables newly added to the config in parallel with a {@link BinlogReader} for all the tables previously
+ * in the config.
+ */
+public class ParallelSnapshotReader implements Reader {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final BinlogReader oldTablesReader;
+    private final BinlogReader newTablesBinlogReader;
+    private final ChainedReader newTablesReader;
+
+    private final AtomicBoolean running = new AtomicBoolean();
+    private final AtomicReference<Runnable> uponCompletion = new AtomicReference<>();
+
+    /**
+     * Create a ParallelSnapshotReader.
+     *
+     * @param config the current connector configuration.
+     * @param noSnapshotContext The context for those tables not undergoing a snapshot.
+     * @param snapshotFilters {@link Filters} matching the tables that should be snapshotted.
+     */
+    public ParallelSnapshotReader(Configuration config,
+                                  MySqlTaskContext noSnapshotContext,
+                                  Filters snapshotFilters) {
+        AtomicBoolean oldTablesReaderNearEnd = new AtomicBoolean(false);
+        AtomicBoolean newTablesReaderNearEnd = new AtomicBoolean(false);
+        ParallelHaltingPredicate oldTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(oldTablesReaderNearEnd, newTablesReaderNearEnd);
+        ParallelHaltingPredicate newTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(newTablesReaderNearEnd, oldTablesReaderNearEnd);
+
+        this.oldTablesReader = new BinlogReader("oldBinlog", noSnapshotContext, oldTablesReaderHaltingPredicate);
+
+        MySqlTaskContext newTablesContext = new MySqlTaskContext(config, snapshotFilters);
+        SnapshotReader newTablesSnapshotReader = new SnapshotReader("newSnapshot", newTablesContext);
+
+        this.newTablesBinlogReader = new BinlogReader("newBinlog", newTablesContext, newTablesReaderHaltingPredicate);
+        this.newTablesReader = new ChainedReader().add(newTablesSnapshotReader).add(newTablesBinlogReader);
```

Comment:
Not related to this PR directly, but I think we should get rid of `ChainedReader#add()` and make it immutable. I've logged [DBZ-583](https://issues.jboss.org/browse/DBZ-583) for this.